### PR TITLE
File System Cleanup and Boost Fallback for libEGM

### PR DESCRIPTION
**Pull request by RobertBColton**
Uses 5 random digits (unique to each emake process) prefixed to the string "gmk_data" with a postfix of the increment of the file number being written. Fundies wants the prefix so he can sort with the file manager. This version does not rely on any POSIX functions or non-standard C++ features.

### CODE REVIEW

File: CommandLine/libEGM/filesystem.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#include "filesystem.h"
```
Comment (JoshDreamland) [inline CommandLine/libEGM/filesystem.cpp]: You can inline this file using templates and/or `using`, but it should be fine like this.

File: CommandLine/libEGM/gmk.cpp [hunk 2]
```diff
@@ -44,17 +48,26 @@ ostream err(nullptr);
...
 }
 
 std::string writeTempDataFile(char *bytes, size_t length) {
-  char temp[] = "gmk_data.XXXXXX";
-  int fd = mkstemp(temp);
-  if (fd == -1) return "";
-  write(fd, bytes, length);
-  close(fd);
-  return temp;
+  static int increment = 0;
```
Comment (JoshDreamland) [inline CommandLine/libEGM/gmk.cpp]: Why did you bother doing all the work to extract our generalized filesystem logic from EGM, only to ignore `std::filesystem::temp_directory_path` / `boost::filesystem::unique_path`?

Comment (RobertBColton) [reply CommandLine/libEGM/gmk.cpp]: Addressed: 9df6bd2b608ef5dd59b1c69583c75ddf47fc9f5d

File: CommandLine/libEGM/filesystem.cpp (hunk 1 above)
Comment (RobertBColton) [reply CommandLine/libEGM/filesystem.cpp]: I don't know what you mean.

**Comment by RobertBColton** [comment]
@RemoveRusky the build failing right now about `filesystem.h` is a great example of what happens when you don't `make clean` when rebuilding. I can reproduce it locally by making `libEGM` on this pull request and then `git checkout master` followed by `make libEGM`.

```bash
make[2]: Entering directory '/c/Users/Owner/Desktop/enigma-dev/CommandLine/libEGM'
make[2]: *** No rule to make target 'filesystem.h', needed by '.eobjs/egm.o'.  Stop.
```

Because this pull request adds the header, the `ci-regression.sh` script is hitting this when it goes to check out master.

**Comment by RobertBColton** [comment]
For now, I got around it by just having fundies let me check in `filesystem.h` first in #1364. This should hopefully fix it.

**Pull request merged**